### DISCUSSION TITLE
Issue with migrating JSR223-pre-processors with Java code (lambdas) to jsrPreProcessor with string scripts when running DSL script with Blazemeter

### DISCUSSION THREAD

**Discussion by cspriya** [abstracta/jmeter-java-dsl#196] [Q&A] [answered]
Hi, 
I am new the Java Jmeter DSL and I am facing issue when I am migrating below script which puts newly generated Authorization in ${Authorization} header placeholder of HTTP sampler to jsrPreProcessor with string scripts since LAMDA expression does not work with Blazemeter execution
`.children( jsr223PreProcessor( s ->
                                                                     s.vars.put(
                                                                                "Authorization",
                                                                             getAuthorizationHeader())))`
                                              
I am not sure how I can convert the above code snippet to jsrPreProcessor String script. I tried with the reference
 [provide-request-parameters-programmatically-per-request](https://abstracta.github.io/jmeter-java-dsl/guide/#provide-request-parameters-programmatically-per-request) did something like below. 

`[.children(
                                        jsr223PreProcessor("vars.put('Authorization', " + getClass().getName()
                                                + ".getAuthorizationHeader(vars))")
                                )]`

Getting below error in Blazemeter logs as Authorization value is not setting with above code in Blazemeter 

2023-06-22 10:43:48,941 ERROR o.a.j.m.JSR223PreProcessor: Problem in JSR223 script, JSR223 PreProcessor
javax.script.ScriptException: groovy.lang.MissingPropertyException: No such property: com for class: Script1
	at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:320) ~[groovy-jsr223-3.0.7.jar:3.0.7]
	at org.codehaus.groovy.jsr223.GroovyCompiledScript.eval(GroovyCompiledScript.java:71) ~[groovy-jsr223-3.0.7.jar:3.0.7]
	at javax.script.CompiledScript.eval(CompiledScript.java:92) ~[?:1.8.0_362]
	at org.apache.jmeter.util.JSR223TestElement.processFileOrScript(JSR223TestElement.java:217) ~[ApacheJMeter_core.jar:5.4.3]
	at org.apache.jmeter.modifiers.JSR223PreProcessor.process(JSR223PreProcessor.java:45) [ApacheJMeter_components.jar:5.4.3]
	at org.apache.jmeter.threads.JMeterThread.runPreProcessors(JMeterThread.java:965) [ApacheJMeter_core.jar:5.4.3]
	at org.apache.jmeter.threads.JMeterThread.executeSamplePackage(JMeterThread.java:549) [ApacheJMeter_core.jar:5.4.3]
	at org.apache.jmeter.threads.JMeterThread.processSampler(JMeterThread.java:489) [ApacheJMeter_core.jar:5.4.3]
	at org.apache.jmeter.threads.JMeterThread.run(JMeterThread.java:256) [ApacheJMeter_core.jar:5.4.3]
	at java.lang.Thread.run(Thread.java:750) [?:1.8.0_362]

Please help me with migrating code from java lamda to run the same test in Blazemeter tool also.

Thanks

Please help @rabelenda

**Comment by rabelenda** [comment]
Something you may try is using the java lambda as is and include jmeter-java-dsl jar and a jar containing any classes you may need (for example the class containing the getAuthorizationHeader() method. I haven't tried this myself, but it should work even though would require proper packaging of custom logic (like getAuthorizationHeader).

The usual way I handle such scenarios is just converting the logic to groovy and using groovy script. In general converting logic to groovy is pretty straightforward since groovy is compatible with java and a lot of java code is valid groovy code. But again, depends on what you have inside getAuthorizationHeader. If you use some jar dependency for instance, you would need to add it as asset to run in Blazmeter.

Can you provide more details of getAuthorizationHeader logic?

In the future would be nice to have automatic lambda execution support in BlazeMeter (or any other cloud provider). We haven't invested on this, but if more people are interested in such support we me reprioritize.

**Reply by cspriya** [reply]
Hi Thanks for quick response, It would be really useful if you can provide lambda support in BlazeMeter. It is not working with the "vars.put('Authorization', " + getClass().getName() + ".getAuthorizationHeader(vars))")  this way as it return the object of getAuthorizationHeader() method. This method is in the same class where the @test method is written.

```java
public static String getAuthorizationHeader(JMeterVariables vars)
            throws URISyntaxException, IOException, UnrecoverableKeyException, CertificateException,
                    KeyStoreException, NoSuchAlgorithmException {
        URI uri = new URIBuilder(Constants.CCA_API_URL).build();
        Charset charset = StandardCharsets.UTF_8;
        PrivateKey signingKey =
                AuthenticationUtils.loadSigningKey(
                        new ClassPathResource(Constants.ACCESS_PATH_8809)
                                .getFile()
                                .getAbsolutePath(),
                        Constants.SIGNING_KEY_ALIAS_8809,
                        Constants.SIGNING_KEY_PASSWORD);
        String authHeader =
                OAuth.getAuthorizationHeader(
                        uri,
                        "POST",
                        requestData,
                        (Charset) charset,
                        Constants.CLIENT_ID_8809,
                        signingKey);
        return authHeader;
    }

  
    @Test
    public void testPerformance() throws IOException, InterruptedException, TimeoutException, UnrecoverableKeyException, CertificateException, URISyntaxException, KeyStoreException, NoSuchAlgorithmException {
        TestPlanStats stats = testPlan(
                threadGroup(1, 1,
                        httpSampler(Constants.CCA_API_URL)
                                .header("Authorization", "${Authorization}")
                                .post(requestData, ContentType.APPLICATION_JSON)
                                .children(
                                        jsr223PreProcessor("vars.put('Authorization', " + getClass().getName()
                                                + ".getAuthorizationHeader(vars))")
                                )
                        //htmlReporter(Constants.reportsDir + "performance-REPORT")
                )
               // resultsTreeVisualizer()
        ).runIn(new BlazeMeterEngine(BlazemeterAPIAuth)
                .testName("Test DSL")
                .totalUsers(1)
                .holdFor(Duration.ofMinutes(5))
                .threadsPerEngine(1)
                .testTimeout(Duration.ofMinutes(5)));
        assertThat(stats.overall().errorsCount()).isEqualTo(0);
        assertThat(stats.overall().sampleTimePercentile99()).isLessThan(Duration.ofSeconds(5));
    }
``` 
And also I am not sure how I can write groovy script inside the preprocessor logic as I need this getAuthorizationHeader() method to be called before each sample request hit as it should be unique for each request.

**Reply by rabelenda** [reply]
I see you have dependencies (external jars or classes, like OAuth and AuthenticationUtils) which will need to be uploaded to blazemeter for proper execution.

I just tried following example and worked, maybe you can do something similar:

```xml
<project xmlns="http://maven.apache.org/POM/4.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>us.abstracta.jmeter</groupId>
  <artifactId>jmeter-java-dsl-sample</artifactId>
  <version>0.1-SNAPSHOT</version>

  <properties>
    <!-- This is important since BlazeMeter engine I used at least uses java 8 -->
    <maven.compiler.source>1.8</maven.compiler.source>
    <maven.compiler.target>1.8</maven.compiler.target>
  </properties>

  <dependencies>
    <dependency>
      <groupId>us.abstracta.jmeter</groupId>
      <artifactId>jmeter-java-dsl</artifactId>
      <version>1.12</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>us.abstracta.jmeter</groupId>
      <artifactId>jmeter-java-dsl-blazemeter</artifactId>
      <version>1.12</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-engine</artifactId>
      <version>5.9.0</version>
      <scope>test</scope>
    </dependency>
  </dependencies>

  <build>
    <pluginManagement>
      <plugins>
        <plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-surefire-plugin</artifactId>
          <version>3.0.0-M7</version>
        </plugin>
        <plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-compiler-plugin</artifactId>
          <version>3.10.1</version>
        </plugin>
      </plugins>
    </pluginManagement>
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-failsafe-plugin</artifactId>
        <version>3.0.0-M7</version>
        <executions>
          <execution>
            <goals>
              <goal>integration-test</goal>
              <goal>verify</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-jar-plugin</artifactId>
        <version>3.3.0</version>
        <executions>
          <execution>
            <goals>
              <goal>test-jar</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-dependency-plugin</artifactId>
        <version>3.6.0</version>
        <executions>
          <execution>
            <id>copy-dependencies</id>
            <phase>package</phase>
            <goals>
              <goal>copy-dependencies</goal>
            </goals>
            <configuration>
              <outputDirectory>${project.build.directory}/libs</outputDirectory>
              <!-- include here any additional dependencies you need to upload to BlazeMeter -->
              <includeArtifactIds>jmeter-java-dsl</includeArtifactIds>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>
</project>
```

```java
package us.abstracta.jmeter.javadsl.sample;

import static us.abstracta.jmeter.javadsl.JmeterDsl.*;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.apache.jmeter.samplers.SampleResult;
import org.junit.jupiter.api.Test;
import us.abstracta.jmeter.javadsl.blazemeter.BlazeMeterEngine;

public class PerformanceIT {

  public static void process(SampleResult sampleResult) {
    sampleResult.setResponseMessage("SUPER");
  }

  @Test
  public void testPerformance() throws Exception {
    testPlan(
        threadGroup(1, 1,
            jsr223Sampler(getClass().getName() + ".process(SampleResult)")
        )
    ).runIn(new BlazeMeterEngine(System.getenv("BZ_TOKEN"))
        .assets(findAssets().toArray(new File[0])));
  }

  private List<File> findAssets() {
    List<File> ret = new ArrayList<>();
    ret.add(new File("target/jmeter-java-dsl-sample-0.1-SNAPSHOT-tests.jar"));
    ret.addAll(Arrays.asList(new File("target/libs").listFiles()));
    return ret;
  }

}
```

**Reply by cspriya** [reply]
I didnt quite get the solution here. Can you please explain what we are doing here. 

I did not find the add asset method in Blazemeter latest version  which you have given in the solution
.runIn(new BlazeMeterEngine(System.getenv("BZ_TOKEN"))
        **.assets(findAssets().toArray(new File[0])));**

 What I am understanding here is I just have to add the Oauth dependency again like below
 ```
<plugin>
                    <groupId>org.apache.maven.plugins</groupId>
                    <artifactId>maven-dependency-plugin</artifactId>
                    <version>3.6.0</version>
                    <executions>
                        <execution>
                            <id>copy-dependencies</id>
                            <phase>package</phase>
                            <goals>
                                <goal>copy-dependencies</goal>
                            </goals>
                            <configuration>
                                <outputDirectory>${project.build.directory}/libs</outputDirectory>
                                <dependency>
                                    <groupId>org-dep</groupId>
                                    <artifactId>oauth1-signer</artifactId>
                                    <version>1.5.1</version>
                                </dependency>
                                <includeArtifactIds>jmeter-java-dsl</includeArtifactIds>
                            </configuration>
                        </execution>
                    </executions>
                </plugin>
```

**Comment by rabelenda** [comment]
Hello, basically in the pom configured:
* failsafe plugin to run integration tests (which run after packaging)
* jar plugin to generate a jar including tests code
* dependencies plugin to copy selected project dependencies to target/lib folder (you should add any other potential dependencies that you need here, for example: Oauth library).

The code just defines the jars (test jar and dependencies) as assets to be uploaded to BlazeMeter test.

The assets method is available since version 1.11. You can check [here](https://github.com/abstracta/jmeter-java-dsl/blob/7fdd59477ad60d59deb91873c63f4c03b0ef486f/jmeter-java-dsl-blazemeter/src/main/java/us/abstracta/jmeter/javadsl/blazemeter/BlazeMeterEngine.java#L294). Maybe you missed updating your IDE pom or something so it finds latest methods (in IntelliJ at least, every time you update the pom, you need to hit a little button that appears with an m and two curved arrows)?

**Reply by rabelenda** [reply]
To add the dependency you just add it to includeArtifactIds property, separated by commas and only including the id. Eg:

```
<includeArtifactIds>jmeter-java-dsl,oauth1-signer</includeArtifactIds>
```

It gets the rest of the info of the artifact from the dependencies info.

**Reply by cspriya** [reply]
Hi I did the same and added the jar and used with blazemeter engine but while executing the the groovy command I am getting this error 

`.children( jsr223PreProcessor("vars.put('Authorization',"+ Thread.currentThread().getClass().getName() + ".getAuthorizationHeader())"))     `                           

javax.script.ScriptException: groovy.lang.MissingMethodException: No signature of method: static java.lang.Thread.getAuthorizationHeader() is applicable for argument types: () values: []

What is wrong in this code?

**Reply by rabelenda** [reply]
`Thread.currrentThread().getClass()` gets the class of the current thread, which is `Thread`. I guess you want `this.getClass()` or something similar.

**Reply by cspriya** [reply]
I have tried all the ways using getClass().getName AND this.getClass(),getName() AND this.getClass(),, For all the cases getting below exception :-

`2023-06-26 06:33:27,113 ERROR o.a.j.m.JSR223PreProcessor: Problem in JSR223 script, JSR223 PreProcessor javax.script.ScriptException: groovy.lang.MissingPropertyException: No such property: com for class: Script1`

**Reply by rabelenda** [reply]
Hello,

You only get this error when running in BlazeMeter? Does it work locally?

If it doesn't work locally, have you tried showInGUI, to check the groovy you generate is actually the one you expect?

If the problem is only in BlazeMeter, have you checked that the testplan has proper jars uploaded? You may even check if the jars actually contain the classes you need (unzipping the jar).

**Reply by rabelenda** [reply]
If you need more real time help, and potentially private channel, you may use [discord](https://discord.com/invite/WNSn5hqmSd) and contact me there. Regards

**Reply by rabelenda** [reply]
Hello @cspriya, any updates on this?

**Reply by cspriya** [reply]
Hi @rabelenda  sorry for late reply , I have tried sending groovy script as file and the file has completer code , that is working fine

.children(jsr223PreProcessor(new String(Files.readAllBytes(Paths.get(
                        new ClassPathResource(data.get("authGroovyScriptPath"))
                                .getFile()
                                .getAbsolutePath())))))))

**Reply by rabelenda** [reply]
So I guess you are no longer using getAuthorizationHeader method in test. Is good that you found a solution, but I think the method in test is simpler to maintain that groovy script (in case you want to try again fixing such approach).

In the solution I provided there are some key points that I didn't describe earlier:

* maven-jar-plugin with test-jar goal: generates a jar with tests code. This is necessary to later upload the test code to BlazeMeter as to be available for execution.
* maven-failsafe-plugin: allows to run tests after the jars are generated (with mvn clean install or mvn clean verify, or doing a mvn package and then running test with IDE). By default, it runs all tests ending it IT. This allows for the test class to reference it's own jar and use it in findAssets, so the code containing the method is properly uploaded to BlazeMeter.
* test class name ending in IT: as for maven-failsafe-plugin to run it after the test jar is generated, and being able to upload the test code to BlazeMeter.
* findAssets in test class referencing test jar: to upload it to BlazeMeter.

Were you aware of this?

**Reply by rabelenda** [reply, marked as answer]
We have released a [new version](https://github.com/abstracta/jmeter-java-dsl/releases/tag/v1.14) which provides an alternative for running lambda code in BlazeMeter and a [new user guide section](https://abstracta.github.io/jmeter-java-dsl/guide/#lambdas) with details.
